src/orion/algo/registry.py

```diff
@@ -0,0 +1,147 @@
+""" Classes that serve as an in-memory storage of trials for the algorithms. """
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from logging import getLogger as get_logger
+from typing import Any, Container, Iterator, Mapping
+
+from orion.core.worker.trial import Trial, TrialCM
+
+logger = get_logger(__name__)
+
+
+class Registry(Container[Trial]):
+    """In-memory container for the trials that the algorithm suggests/observes/etc."""
+
+    def __init__(self):
+        self._trials: dict[str, Trial] = {}
+
+    def __contains__(self, trial_or_id: str | Trial | Any) -> bool:
+        if isinstance(trial_or_id, TrialCM):
+            trial_id = _get_id(trial_or_id._cm_trial)
+        elif isinstance(trial_or_id, Trial):
+            trial_id = _get_id(trial_or_id)
+        elif isinstance(trial_or_id, str):
+            trial_id = trial_or_id
+        else:
+            raise NotImplementedError(trial_or_id)
+        return trial_id in self._trials
+
+    def __getitem__(self, item: int | str) -> Trial:
+        if not isinstance(item, (str, int)):
+            raise KeyError(item)
+        if isinstance(item, str):
+            return self._trials[item]
+        trial_ids = list(self._trials.keys())
+        return self._trials[trial_ids[item]]
+
+    def __iter__(self) -> Iterator[Trial]:
+        return iter(self._trials.values())
+
+    def __len__(self) -> int:
+        return len(self._trials)
+
+    @property
+    def state_dict(self) -> dict:
+        return {"_trials": self._trials}
```

Comment:
Maybe we should deepcopy here and in set_state to be safe.

Reply:
Fixed in eaa2254453102689c230a4b5fb12952b9be77b20 (not pushed yet)

Reply:
Commit is for hyperband.

Reply:
Oops, wrong commit, but it's fixed: https://github.com/lebrice/orion/blob/eebe53ca24d35db5b9cb917c3a319e4755d41266/src/orion/algo/registry.py#L45

Reply:
And set_state?https://github.com/lebrice/orion/blob/eebe53ca24d35db5b9cb917c3a319e4755d41266/src/orion/algo/registry.py#L49

Reply:
Fixed in 15ea577e2f3f8f8407a4a311d2e4d39f8e11056e (not pushed atm)